contracts/src/StakeTable.sol

```diff
@@ -52,493 +102,356 @@ contract StakeTable is AbstractStakeTable, Ownable, InitializedAt {
-    /// @dev This function can only be called by the validator itself when it's not in the exit
-    /// queue
+    /// @dev This function can only be called by the validator itself when it hasn't exited
+    ///      TODO: MA: is this a good idea? Why should key rotation be blocked for an exiting
+    ///      validator?
     /// @dev The validator will need to give up either its old BLS key and/or old Schnorr key
     /// @dev The validator will need to provide a BLS signature to prove that the account owns the
     /// new BLS key
     /// @param newBlsVK The new BLS verification key
     /// @param newSchnorrVK The new Schnorr verification key
     /// @param newBlsSig The BLS signature that the account owns the new BLS key
+    ///
+    /// TODO: MA: I think this function should be reworked. Is it fine to always force updating both
+    /// keys? If not we should probably rather have two functions for updating the keys. But this
+    /// would also mean two separate events, or storing the keys in the contract only for this
+    /// update function to remit the old keys, or throw errors if the keys are not changed. None of
+    /// that seems useful enough to warrant the extra complexity in the contract and GCL.
     function updateConsensusKeys(
         BN254.G2Point memory newBlsVK,
         EdOnBN254.EdOnBN254Point memory newSchnorrVK,
         BN254.G1Point memory newBlsSig
-    ) external virtual override {
-        Node memory node = nodes[msg.sender];
-
-        if (node.account == address(0)) revert NodeNotRegistered();
-
-        // Verify that the node is not in the exit queue
-        if (node.exitEpoch != 0) revert ExitRequestInProgress();
-
-        if (_isEqualBlsKey(newBlsVK, node.blsVK) && newSchnorrVK.isEqual(node.schnorrVK)) {
-            revert NoKeyChange();
-        }
+    ) external virtual {
+        ensureValidatorActive(msg.sender);
+        ensureNonZeroSchnorrKey(newSchnorrVK);
+        ensureNewKey(newBlsVK);
 
-        // Zero-point constants for verification
-        BN254.G2Point memory zeroBlsKey = BN254.G2Point(
-            BN254.BaseField.wrap(0),
-            BN254.BaseField.wrap(0),
-            BN254.BaseField.wrap(0),
-            BN254.BaseField.wrap(0)
-        );
-        EdOnBN254.EdOnBN254Point memory zeroSchnorrKey = EdOnBN254.EdOnBN254Point(0, 0);
-
-        if (_isEqualBlsKey(newBlsVK, zeroBlsKey)) revert InvalidBlsVK();
-
-        if (newSchnorrVK.isEqual(zeroSchnorrKey)) revert InvalidSchnorrVK();
-
-        // Verify that the validator can sign for that newBlsVK, otherwise it inner reverts with
-        // BLSSigVerificationFailed
+        // Verify that the validator can sign for that blsVK. This prevents rogue public-key
+        // attacks.
         bytes memory message = abi.encode(msg.sender);
         BLSSig.verifyBlsSig(message, newBlsSig, newBlsVK);
 
-        node.blsVK = newBlsVK;
+        blsKeys[_hashBlsKey(newBlsVK)] = true;
```

Comment:
wondering how are bls keys currently associated with ethereum accounts in the contract? i believe there should be an association as delegations are based on the validator ethereum account address. 

Maybe this is handled in the confirmation layer?

Reply:
Via the event

```
    event ValidatorRegistered(
        address indexed account,
        BN254.G2Point blsVk,
        EdOnBN254.EdOnBN254Point schnorrVk,
        uint16 commission
    );
```

the contract doesn't currently associate them in the contract storage.

Reply:
So the register and key update events tell the confirmation layer what consensus keys to use for which Ethereum account. The users don't really have to worry about the keys of validators although we will have APIs for this in the confirmation layer so that User interfaces can show all the information. The contract is kind of minimal now in terms of what it stores and computes, which the exception of checking that keys are new and the BLS signature check but we might move all of this to the confirmation layer too.

For slashing we will ensure that the address to slash can be authenticated via merkle proof against the LC state root.

Does this answer your concerns?

Reply:
gotcha - and in hotshot they would be keeping track of it so even though the bls key changed, they know it's the same ethereum accoun that owns it? 

that's fair since the user would never care about bls keys

Reply:
it does! thank yout